### PR TITLE
[DRAFT] Split generated files for error and version_features

### PR DESCRIPTION
**Pull request by mpg**
## Description

This PR, based on development-psa, is an exploration of a possible strategy to support deduplicating most files between mbedtls and mbed-crypto without breaking the API as an intermediate step. As such, it is a companion to https://github.com/ARMmbed/mbedtls/pull/2349 and will soon be completed by a technical analysis document.

The key observation is that it's easy to have specific variants of one thing on the TLS side as long as it's in headers. This is thanks to the carefully chosen include order, originally intended for `config.h`, and also used for `version.h` in #2349. We can extend that strategy to the generated files containing error strings rest. feature strings, in order to have only the crypto part on the crypto side, but get the full version when building libmbedcrypto as a submodule of mbedtls.

If the approach is adopted, this PR should be side-ported to mbed-crypto, then both should be merged before rebasing https://github.com/ARMmbed/mbedtls-psa/pull/232 and https://github.com/ARMmbed/mbedtls/pull/2349, making some adaptations in the later (see below) and merging them.

The adaptation in https://github.com/ARMmbed/mbedtls/pull/2349 would likely consist of just appending `<crypto/include/mbedtls/*.h> to the list on line 40 of `scripts/generate_errors.pl` and checking the result.

## Status

**DRAFT** - for discussion only! - Ping @sbutcher-arm and @Patater 

**Things known to be broken for now:**
- might not work as intended when building from the submodule until sideported to mbed-crypto and submodule commit updated on the mbedtls side

**Tested and working so far:**
- `make all test`
- `cmake . && make all test`
- `tests/scripts/check-generated-files.sh` (plus manual negative testing)

### CODE REVIEW

File: include/mbedtls/error_high.h [hunk 1]
```diff
@@ -0,0 +1,375 @@
+/**
+ * \file error_high.h
```
Comment (gilles-peskine-arm) [inline include/mbedtls/error_high.h]: Why are the generated header files (`error_high.h`, `error_low.h`, `error_includes.h`, `version_features.h`) in `include/mbedtls` and not in `library`? When would a user of the library ever use them?

Comment (mpg) [reply include/mbedtls/error_high.h]: As discussed IRL and for reference here, for a PoC I went for the easiest solution, re-using the existing favourable include order that only applies to `include/mbedtls`, but for a final version we could adapt the include path to allow for a private include dir with the same virtualisation property.

**Comment by mpg** [comment]
@Patater I'm not familiar with the Mbed OS importer and the Mbed OS build system, so I forgot to ask during the workshop, but how would that kind of header virtualisation affect it? In particular:

- would it work the same way if the virtualised headers were put in a sub-directory of `library` (or some other non-public location) as @gilles-peskine-arm  suggested?
- what's the interaction with your PR https://github.com/ARMmbed/mbed-os/pull/9463 that allows to import Mbed Crypto separately? Does it need a sort of blacklist of files to be updated only via Mbed TLS and not be overwritten by the Mbed Crypto version?

Thanks in advance for clarifying that!

**Comment by mpg** [comment]
Closing this draft as an actual PR is up with a new, cleaner approach: https://github.com/ARMmbed/mbed-crypto/pull/97

**Pull request closed without merging**